atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionCacheValueStoreImpl.java

```diff
@@ -0,0 +1,152 @@
+            localUpdates.put(cellReference, LocalCacheEntry.write(value));
+        }
+    }
+
+    @Override
+    public void updateLocalReads(TableReference tableReference, Map<Cell, byte[]> remoteReadValues) {
+        if (snapshot.isWatched(tableReference)) {
+            KeyedStream.stream(remoteReadValues)
+                    .mapKeys(cell -> CellReference.of(tableReference, cell))
+                    .map(CacheValue::of)
+                    .filterKeys(snapshot::isUnlocked)
+                    .forEach((cell, value) -> localUpdates.put(cell, LocalCacheEntry.read(value)));
+        }
+    }
+
+    @Override
+    public void updateEmptyReads(TableReference tableReference, Set<CellReference> emptyCells) {
+        if (snapshot.isWatched(tableReference)) {
+            emptyCells.stream()
+                    .filter(snapshot::isUnlocked)
+                    .forEach(cell -> localUpdates.put(cell, LocalCacheEntry.read(CacheValue.empty())));
+        }
+    }
+
+    @Override
+    public Map<CellReference, CacheValue> getCachedValues(Set<CellReference> cellReferences) {
+        Map<CellReference, CacheValue> locallyCachedValues = getLocallyCachedValues(cellReferences);
+
+        // Filter out which values have not been read yet
+        Set<CellReference> remainingCells = Sets.difference(cellReferences, locallyCachedValues.keySet());
+
+        // Read values from the snapshot. For the hits, mark as hit in the local map.
+        Map<CellReference, CacheValue> snapshotCachedValues = getSnapshotValues(remainingCells);
+        snapshotCachedValues.forEach(
+                (cellReference, value) -> localUpdates.put(cellReference, LocalCacheEntry.hit(value)));
+
+        return ImmutableMap.<CellReference, CacheValue>builder()
+                .putAll(locallyCachedValues)
+                .putAll(snapshotCachedValues)
+                .build();
+    }
+
+    @Override
+    public Map<CellReference, CacheValue> getTransactionDigest() {
+        return KeyedStream.stream(localUpdates)
+                .filter(entry -> entry.status().equals(Status.READ))
+                .map(LocalCacheEntry::value)
+                .collectToMap();
+    }
+
+    private Map<CellReference, CacheValue> getLocallyCachedValues(Set<CellReference> cellReferences) {
+        return KeyedStream.of(cellReferences)
+                .map(localUpdates::get)
+                .filter(Objects::nonNull)
+                .map(LocalCacheEntry::value)
+                .collectToMap();
+    }
+
+    private Map<CellReference, CacheValue> getSnapshotValues(Set<CellReference> thusFarUncachedValues) {
+        return KeyedStream.of(thusFarUncachedValues)
+                .map(snapshot::getValue)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .filter(CacheEntry::isUnlocked)
+                .map(CacheEntry::value)
+                .collectToMap();
+    }
+
+    @Value.Immutable
+    public interface LocalCacheEntry {
```

Comment:
No, I don't want to use `derive4j` because I think it's annoying.

Reply:
Yeah, I agree that in those cases, derive4j might be nicer, but here I think not.